tidy3d/components/device/simulation.py

```diff
@@ -327,9 +351,122 @@ def check_medium_heat_spec(cls, values):
+
+        HeatBCTypes = (TemperatureBC, HeatFluxBC, ConvectionBC)
+        HeatSourceTypes = (UniformHeatSource, HeatSource, HeatFromElectricSource)
+        ElectricBCTypes = (VoltageBC, CurrentBC, InsulatingBC)
+
+        sources = values["sources"]
+        monitors = values["monitors"]
+        structures = values["structures"]
+        boundary_specs = values["boundary_spec"]
+        background_medium = values["medium"]
+
+        simulation_types = []
+        # check bcs
+        for bc in boundary_specs:
+            if isinstance(bc.condition, HeatBCTypes):
+                if DeviceSimulationType.HEAT not in simulation_types:
+                    simulation_types.append(DeviceSimulationType.HEAT)
+            elif isinstance(bc.condition, ElectricBCTypes):
+                if DeviceSimulationType.CONDUCTION not in simulation_types:
+                    simulation_types.append(DeviceSimulationType.CONDUCTION)
+
+        # check structures (make sure at least one has appropriate specs)
+        NeutralTypes = (FluidSpec, InsulatorSpec)
+        for st in simulation_types:
+            one_correct = False
+            for struct in structures:
+                medium = struct.medium
+                if st == DeviceSimulationType.HEAT:
+                    if (
+                        not isinstance(medium.heat_spec, NeutralTypes)
+                        and medium.heat_spec is not None
+                    ):
+                        one_correct = True
+                elif st == DeviceSimulationType.CONDUCTION:
+                    if (
+                        not isinstance(medium.electric_spec, NeutralTypes)
+                        and medium.electric_spec is not None
+                    ):
+                        one_correct = True
+
+            # check background medium
+            if st == DeviceSimulationType.HEAT:
+                if (
+                    not isinstance(background_medium.heat_spec, NeutralTypes)
+                    and background_medium.heat_spec is not None
+                ):
+                    one_correct = True
+            elif st == DeviceSimulationType.CONDUCTION:
+                if (
+                    not isinstance(background_medium.electric_spec, NeutralTypes)
+                    and background_medium.electric_spec is not None
+                ):
+                    one_correct = True
+
+            if not one_correct:
+                raise SetupError(
+                    "Make sure your structures have the correct HEAT/CONDUCTION specification, e.g., "
+                    'for simulations of HEAT type, "heat_spec" must be not None.'
+                )
+
+        return values
+
+    @staticmethod
+    def _check_simulation_types(values: Dict) -> Tuple[int, ...]:
+        """Given model dictionary ``values``, check the type of simulations to be run
+        based on BCs and sources.
+        """
+        # NOTE: this function does not consider the source 'HeatFromElectricSource'
```

Comment:
I have changed this so that it can also check for `HeatFromElectricSource`. So far, I wanted to have that flexibility since we're already checking for `HeatFromElectricSource` and associated specs in a different validator.